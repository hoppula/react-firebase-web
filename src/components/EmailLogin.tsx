import * as PropTypes from "prop-types"
import * as React from "react"
import { FirebaseContext } from "../types"

export interface LoginProps {
  readonly children: (
    submit: (email: string, password: string) => Promise<any>
  ) => React.ReactElement<any>
}

export class EmailLogin extends React.Component<LoginProps, {}> {
  static propTypes = {
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  context: FirebaseContext

  submit = (email: string, password: string) => {
    return this.context.firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
  }

  render() {
    return React.Children.only(this.props.children(this.submit))
  }
}

export default EmailLogin
