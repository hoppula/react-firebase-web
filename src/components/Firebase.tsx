import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
