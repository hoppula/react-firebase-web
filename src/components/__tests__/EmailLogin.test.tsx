import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import EmailLogin from "../EmailLogin"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})

describe("EmailLogin", () => {
  test("logs user in using email and password", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <EmailLogin
          children={login => (
            <button onClick={() => login("test@email.dev", "password")}>
              Login
            </button>
          )}
        />
      </Firebase>
    )
    const auth = component.getInstance().firebase.auth()
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(auth.currentUser).toBeNull()

    const button = component.root.findByType("button")
    button.props.onClick()
    auth.flush()

    expect(auth.currentUser.email).toBe("test@email.dev")
  })
})
