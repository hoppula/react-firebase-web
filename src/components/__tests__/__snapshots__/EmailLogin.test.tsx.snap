// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailLogin logs user in using email and password 1`] = `
<button
  onClick={[Function]}
>
  Login
</button>
`;
